src/malcolmWidgets/table/virtualizedTable.component.js

```diff
@@ -0,0 +1,374 @@
+    backgroundColor: theme.palette.background.default,
+    textAlign: 'Center',
+    padding: '2px',
+  },
+  button: {
+    width: '24px',
+    height: '24px',
+    '&:hover': {
+      backgroundColor: 'transparent',
+    },
+  },
+});
+
+const widgetWidths = {
+  'widget:checkbox': 48,
+  'widget:led': 36,
+  'info:alarm': 36,
+};
+
+const getTableState = props => {
+  const tableState = {};
+  tableState.columnLabels =
+    props.localState === undefined
+      ? Object.keys(props.attribute.raw.meta.elements)
+      : props.localState.labels;
+  tableState.values =
+    props.localState === undefined
+      ? props.attribute.raw.value[tableState.columnLabels[0]].map(
+          (val, row) => {
+            const rowData = {};
+            tableState.columnLabels.forEach(label => {
+              rowData[label] = props.attribute.raw.value[label][row];
+            });
+            return rowData;
+          }
+        )
+      : props.localState.value;
+  tableState.flags =
+    props.localState === undefined
+      ? { rows: [], table: {} }
+      : props.localState.flags;
+  tableState.meta =
+    props.localState === undefined
+      ? props.attribute.raw.meta
+      : props.localState.meta;
+  return tableState;
+};
+
+const getWidgetWidth = tag => widgetWidths[tag];
+
+const getColumnWidths = (
+  divWidth,
+  columnWidgetTags,
+  columnLabels,
+  hideInfo
+) => {
+  const fixedWidths = columnWidgetTags.map(
+    (tag, index) =>
+      getWidgetWidth(tag) !== undefined
+        ? Math.max(getWidgetWidth(tag), 11 * columnLabels[index].length + 4)
+        : undefined
+  );
+  const usedWidth = fixedWidths
+    .filter(val => val !== undefined)
+    .reduce((total, val) => total + val);
+  const numVariableWidth = fixedWidths.filter(val => val === undefined).length;
+  return divWidth - usedWidth - 36 * !hideInfo > numVariableWidth * 180
+    ? fixedWidths.map(
+        val =>
+          val !== undefined
+            ? val
+            : (divWidth - usedWidth - 36 * !hideInfo) / numVariableWidth
+      )
+    : fixedWidths.map(val => (val !== undefined ? val : 180));
+};
+
+const isSelectedRow = (row, props) => {
+  const classes = styles(props.theme);
+  return props.selectedRow === row ? classes.selectedRow : classes.textBody;
+};
+const AlarmCell = props => (
+  <IconButton
+    className={props.classes.button}
+    disableRipple
+    onClick={() => {
+      props.infoClickHandler(props.path, `row.${props.row}`);
+    }}
+  >
+    <AttributeAlarm
+      alarmSeverity={
+        props.flags.rows[props.row] &&
+        (props.flags.rows[props.row]._dirty ||
+          props.flags.rows[props.row]._isChanged)
+          ? AlarmStates.DIRTY
+          : AlarmStates.NO_ALARM
+      }
+    />
+  </IconButton>
+);
+
+const getRowData = (row, widgetTags, props, tableState, handlers) => {
+  let valueCells;
+  if (tableState.columnLabels === undefined) {
+    valueCells = {
+      value:
+        row >= tableState.values.length ? null : (
```

Comment:
it feels like lines 135-172 are quite similar, we should probably look at combining if possible in the next update.